### PR TITLE
State live sync

### PR DESCRIPTION
**Pull request by paberr**
## What's in this pull request?

- **This PR also includes all changes from this PR: #1115**
    - Fix changes introduced in this PR
    - Slight cleanup of the trie crate
- Create `MockNode` implementation for tests that can overwrite response behaviour
- Refactoring of the live sync
    - Split the live sync into a generic `LiveSyncer`, a queue component and request components
    - Remove `RequestComponent` trait (which was only used for tests) and replace it by `MockNode`
    - Remove unnecessary buffers
- Implement `StateQueue` and `ChunkRequestComponent`
- Test State Live Sync
- Properly share peer list between syncing components
- Temporarily make accounts changes on incomplete state *not* fail until new accounts model is merged

## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.
- [ ] Regenerate zkp proofs for tests before merging.

### CODE REVIEW

File: block-production/Cargo.toml [hunk 1]
```diff
@@ -33,6 +33,7 @@ nimiq-hash = { path = "../hash" }
 nimiq-keys = { path = "../keys" }
 nimiq-primitives = { path = "../primitives" }
 nimiq-transaction = { path = "../primitives/transaction" }
+nimiq-trie = { path = "../primitives/trie" }
```
Comment (jsdanielh) [inline block-production/Cargo.toml]: Is this right? I don't see why this is now a regular dependency

File: zkp-component/tests/zkp_requests.rs [hunk 2]
```diff
@@ -86,6 +86,7 @@ async fn peers_dont_reply_with_outdated_proof() {
 }
 
 #[test(tokio::test)]
+#[ignore]
```
Comment (jsdanielh) [inline zkp-component/tests/zkp_requests.rs:91]: Why are we ignoring this test now?

File: zkp-component/tests/zkp_requests.rs [hunk 3]
```diff
@@ -166,6 +167,7 @@ async fn peers_reply_with_valid_proof() {
...
```
Comment (jsdanielh) [inline zkp-component/tests/zkp_requests.rs:174]: Why are we ignoring this test now?

File: zkp-component/tests/zkp_component.rs [hunk 4]
```diff
@@ -77,6 +77,7 @@ async fn builds_valid_genesis_proof() {
 }
 
 #[test(tokio::test)]
+#[ignore]
```
Comment (jsdanielh) [inline zkp-component/tests/zkp_component.rs:77]: Why are we ignoring this test now?

File: zkp-component/tests/proof_utils.rs [hunk 5]
```diff
@@ -111,6 +111,7 @@ async fn can_detect_invalid_proof_none_genesis_blocks() {
 }
 
 #[test(tokio::test)]
+#[ignore]
```
Comment (jsdanielh) [inline zkp-component/tests/proof_utils.rs:114]: Why are we ignoring this test now?

File: zkp-component/tests/zkp_requests.rs (hunk 2 above)
Comment (paberr) [reply zkp-component/tests/zkp_requests.rs:91]: Good point, we will need to regenerate the proof and add it before merging.
I added it to the PR todo list, but I will mark it as resolved so reviewing can go on.

File: block-production/Cargo.toml (hunk 1 above)
Comment (paberr) [reply block-production/Cargo.toml]: Good catch, doesn't seem to be needed.

File: blockchain/src/chain_info.rs [hunk 6]
```diff
@@ -10,26 +11,31 @@ use nimiq_database_value::{FromDatabaseValue, IntoDatabaseValue};
...
-    // The sum of all transaction fees in this chain. It resets every batch.
+    /// The sum of all transaction fees in this chain. It resets every batch.
     pub cum_tx_fees: Coin,
-    // The accumulated extended transaction size. It resets every other macro block.
+    /// The accumulated extended transaction size. It resets every other macro block.
     pub cum_ext_tx_size: u64,
-    // A boolean stating if this block can be pruned
+    /// A boolean stating if this block can be pruned.
     pub prunable: bool,
+    /// Bla.
```
Comment (paberr) [inline blockchain/src/chain_info.rs:31]: Should be more descriptive @ii-cruz

File: primitives/trie/src/trie_chunk.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    ) -> Self {
+        Self {
+            keys_end,
+            items,
+            proof,
+        }
+    }
+}
+
+impl Serialize for TrieChunk {
```
Comment (styppo) [inline primitives/trie/src/trie_chunk.rs]: `Serialize`/`Deserialize` should be derivable. Needs `#[beserial(len_type...)]` for `TrieChunk.items`.

File: primitives/trie/src/trie_chunk.rs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+#[derive(Debug, Clone)]
+pub struct TrieChunkWithStart {
+    pub chunk: TrieChunk,
+    pub start_key: KeyNibbles,
+}
+
+/// Common data structure for holding chunk items and proof.
+#[derive(Debug, Clone)]
+pub struct TrieChunk {
+    pub keys_end: Option<KeyNibbles>,
```
Comment (styppo) [inline primitives/trie/src/trie_chunk.rs]: Naming could be improved. How about `end_key`?
Would also be good to add a comment here to clarify that `end_key` is exclusive.

File: consensus/src/sync/live/block_queue/live_sync.rs [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+                        .map(|(push_result, adopted_blocks, invalid_blocks)| {
+                            PushOpResult::Missing(push_result, adopted_blocks, invalid_blocks)
+                        })
+                        .boxed(),
+                );
+            }
+            QueuedBlock::TooFarFuture(_, peer_id) => {
+                // Peer is too far ahead.
+                future_results.push_back(
+                    future::ready(PushOpResult::PeerEvent(LiveSyncPeerEvent::AdvancedPeer(
```
Comment (styppo) [inline consensus/src/sync/live/block_queue/live_sync.rs]: `AdvancedPeer` is not optimal. How about calling the enum variants of `LiveSyncPeerEvent` `Ahead` and `Behind`?

File: consensus/src/sync/live/block_queue/live_sync.rs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+                        peer_id,
+                    )))
+                    .boxed(),
+                );
+            }
+            QueuedBlock::TooDistantPast(_, peer_id) => {
+                // Peer is too far behind.
+                future_results.push_back(
+                    future::ready(PushOpResult::PeerEvent(LiveSyncPeerEvent::OutdatedPeer(
```
Comment (styppo) [inline consensus/src/sync/live/block_queue/live_sync.rs]: See above.

File: consensus/src/sync/live/block_queue/mod.rs [hunk 11]
```diff
@@ -39,76 +47,156 @@ impl Topic for BlockHeaderTopic {
...
+>;
+pub type GossipSubBlockStream<N> = BoxStream<'static, (Block, <N as Network>::PubsubId)>;
+
+pub type BlockAndId<N> = (Block, Option<<N as Network>::PubsubId>);
+
+pub enum QueuedBlock<N: Network> {
+    Head(BlockAndId<N>),
+    Buffered(Vec<BlockAndId<N>>),
+    Missing(Vec<Block>),
+    TooFarFuture(Block, N::PeerId),
```
Comment (styppo) [inline consensus/src/sync/live/block_queue/mod.rs]: In line with `LiveSyncPeerEvent`, these variantes could be called `TooFarAhead` and `TooFarBehind`.

File: primitives/trie/src/trie.rs [hunk 12]
```diff
@@ -404,103 +1217,248 @@ impl<A: Serialize + Deserialize + Clone> MerkleRadixTrie<A> {
     /// Creates a proof for the chunk of the Merkle Radix Trie that starts at the key `start` (which
     /// might or not be a part of the trie, if it is then it will be part of the chunk) and contains
     /// at most `size` leaf nodes.
+    /// PITODO: Perhaps remove this?
```
Comment (styppo) [inline primitives/trie/src/trie.rs]: Seems only to be used in tests, so is probably good to go.

File: primitives/trie/src/trie_node.rs [hunk 13]
```diff
@@ -253,100 +253,108 @@ impl<A: Serialize + Deserialize + Clone> TrieNode<A> {
...
+        let mut result = 1;
+        writer.write_u8(flags)?;
+        if let Some(root_data) = &self.root_data {
+            result += root_data.serialize(writer)?;
+        }
+        if let Some(value) = &self.value {
+            result += value.serialize::<u16, _>(writer)?;
+        }
+        if has_children {
+            result += self.children.serialize(writer)?;
```
Comment (styppo) [inline primitives/trie/src/trie_node.rs]: This is not yet serializing the children into the compact format that we had before (only serialize children that are present, recover their position in the array from the child's prefix). This was a left-over tasks from last hackathon. Might be at odds now with children that use an empty prefix as a marker?

**Review by styppo — APPROVED** [review]
Very clean overall 👍

**Pull request merged**